### PR TITLE
include filter function derivatives developed by Isabel Le and Julian…

### PR DESCRIPTION
**Pull request by JDTeske**
Include filter function gradients implemented by Isabel Le under the supervision of Julian Teske. Refactored by Julian Teske

### CODE REVIEW

**Comment by thangleiter** [comment]
Looks great! I have a few high-level questions and suggestions:

1. Do we really need to cache the derivatives in the `PulseSequence` object? You only need to evaluate the derivative once during optimization, right? As far as I can imagine, only if one wants to compute the derivative of the concatenation of gate operations caching is really required.
2. I implemented caching of intermediate elements of the computation of the control matrix (see [here](https://github.com/qutech/filter_functions/tree/feature/cache_intermediates)) for calculating second order terms. I think this is useful here as well since they are required for the derivative as well (in `gradient.control_matrix_at_timestep_derivative`) and here they are computed twice. 
3. Why pin `sparse<=0.11` in `setup.py`?

**Comment by JDTeske** [comment]
Dear honorable Mr. Hangleiter,
thank you for your positive feedback. Let me answer these questions in utter most brevity.

1. No, its sufficient for my optimizers if the gradients are returned ones. You need to calculate the derivative in every iteration of the optimization. Yes that would be the only use case, I can think of.
-> suggestion: store gradient flag as class attribute in PulseSequence, which is by default set to FALSE
2. Sounds legit. I will investigate this issue ASAP! Are you refering to parts which you only implemented recently?
3. If this was implemented by my pull request then not by purpose.

What is the suggested workflow, if we imply changes? Will you outright reject this pull request and I file another one?

**Comment by thangleiter** [comment]
> suggestion: store gradient flag as class attribute in PulseSequence, which is by default set to FALSE

What should the flag signify?

> Sounds legit. I will investigate this issue ASAP! Are you refering to parts which you only implemented recently?

Yes, see https://github.com/qutech/filter_functions/pull/39

> If this was implemented by my pull request then not by purpose.

Ok, then we can just take that out.

> What is the suggested workflow, if we imply changes? Will you outright reject this pull request and I file another one?

You can just add commits on your remote. This PR will be automatically updated.

File: README.md [hunk 1]
```diff
@@ -47,14 +47,14 @@ Y2.cache_filter_function(omega)
 X.cache_filter_function(omega)
 
 hadamard = Y2 @ X           # equivalent: ff.concatenate((Y2, X))
-hadamard.is_cached('filter function')
+hadamard.is_cached('F')
```
Comment (thangleiter) [inline README.md:50]: This change should be undone (I think something went wrong with merging here).

File: README.md [hunk 2]
```diff
@@ -47,14 +47,14 @@ Y2.cache_filter_function(omega)
...
 # True  (filter function cached during concatenation)
 ```
 
 To compute, for example, the infidelity of the gate in the presence of an arbitrary classical noise spectrum, we can simply call `infidelity()`:
 
 ```python
-spectrum = 1e-2/abs(omega)
+spectrum = 1e-2/abs(omega)  # omega is symmetric about zero
```
Comment (thangleiter) [inline README.md:57]: Same here.

File: filter_functions/gradient.py [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+    n_ops = len(pulse.n_opers)
+    S = np.asarray(S)
+    if(S.shape[0] == n_ops):
+        S_all = S
+    elif(S.shape[0] == 1):
+        S_all = np.array([S[0]]*n_ops)
+    elif(S.shape[0] == len(omega)):
+        S_all = np.array([S]*n_ops)
+    else:
+        raise ValueError('Not fitting shape of S.')
```
Comment (thangleiter) [inline filter_functions/gradient.py:645]: This is functionality already provided by `numeric._get_integrand()`.

File: filter_functions/gradient.py [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+
+    deriv_F = pulse.get_filter_function_derivative(
+        omega=omega, contorl_identifier=control_identifiers, s_derivs=s_derivs)
+    d = pulse.d
+    integrand = np.einsum('ao,atho->atho', S_all, deriv_F)
```
Comment (thangleiter) [inline filter_functions/gradient.py:650]: We should be able to use `numeric._get_integrand` here.

**Review by thangleiter — CHANGES_REQUESTED** [review]

File: filter_functions/gradient.py (hunk 3 above)
Comment (thangleiter) [reply filter_functions/gradient.py:645]: Can be done at a later point when cleaning up. Not worth the effort for now.

**Pull request merged**